probe-rs/src/debug/variable.rs

```diff
@@ -4,36 +4,69 @@ use anyhow::anyhow;
-/// - `<stack_frame>`: Every `StackFrame` will have one of these, with its function name captured in the `value` field of this dummy variable.
-///   - `<statics>`: The parent variable for all statics variables that are in the same compile unit, or in dependencies that are explicitly mentioned in the compile unit of the relevant stack frame.
-///     - A recursive `Variable` structure as described for `<locals>` below.
-///   - `<registers>`: Every `StackFrame` will have a collection of `Variable` registers.
-///     - A `Variable` for each available register
-///   - `<locals>`: Every `StackFrame` (function) will have a collection of locally scoped `Variable`s.
-///     - A `Variable` for each in-scope variable. Complex variables and pointers will have additional children.
-///       - Child `Variable`s that make up a complex parent variable.
-///         - This structure is recursive until a base type is encountered.
+/// We use 'special' `Variables`, to create a structure that models the tree strucutre used by MS DAP specification ( `Threads -> StackTrace -> Scopes -> Variables`), and later on allows the debugger to request these as parent-children structrues.)
+///
+/// In probe-rs, this `Variable` structure looks as follows (indentation represent parent-child structure)
+/// - [VariableName::CoreId] : Every [`Core::id()`] is represented as a top level thread. e.g. Core 0 is thread 0, and so on. This is used as if it is synonomous to the `Threads` level in the MS DAP specification.
+///   - [VariableName::StackFrame]: Every `StackFrame` in a stack trace will have one of these, with its function name captured in the `value` field of this variable.
+///     - [VariableName::StaticScope]: The parent variable for all static scoped variables that are in the same compile unit, or in dependencies that are explicitly mentioned in the compile unit of the relevant stack frame.
+///       - A recursive `Variable` structure as described for `[VariableName::LocalScope]` below.
+///     - [VariableName::Registers]: Every `StackFrame` will have a collection of `Variable` registers with values scoped to that frame.
+///       - A `Variable` for each available register
+///     - [VariableName::LocalScope]: Every `StackFrame` (function) will have a collection of locally scoped `Variable`s.
+///       - A `Variable` for each in-scope variable. Complex variables and pointers will have additional children.
+///         - Child `Variable`s that make up a complex parent variable.
+///           - This structure is recursive until a base type is encountered.
 #[derive(Debug)]
 pub struct VariableCache {
     variable_cache_key: i64,
     variable_hash_map: HashMap<i64, Variable>,
 }
 impl Default for VariableCache {
     fn default() -> Self {
-        Self::new()
+        Self::new(0)
     }
 }
 
 impl VariableCache {
-    pub fn new() -> Self {
-        Self {
+    pub fn new(_core_index: usize) -> Self {
```

Comment:
The `core_index` becomes the `Variable::variable_key` of the `VariableName::CoreId` variable. The intent is that later in code, that allows you to retrieve the variables belonging to the specific `core_index`. As I proceed with the WIP, if this turns out to be unnecessary, I will remove it.

Reply:
As I understand it, there can be stacktraces for multiple cores in a single `VariableCache`. What does the `core_index` mean in that case?

Reply:
Let me finish the next part of the WIP code, hopefully that will make it more clear ... but in a nutshell, the `core_idx` will be used to create a `VariableName::CoreId` variable, that is the parent of all stack trace artifacts (stackframes, scopes, variables) for that core.

Reply:
PS. I just realized that why this might be confusing ...  Are you asking because this only allows one `core_index` per `VariableCache`?

Reply:
> PS. I just realized why this might be confusing ... Are you asking because this only allows one `core_index` per `VariableCache` in the `new()`, when we said the cache can hold data for multiple cores?

Yes, that this what my question is about? :)

Reply:
Sorry about being slow to catch on about that  :).   The mismatch is a direct result of this being WIP code, I should have made that more clear in the commit I sent you. Hopefully the next commit will clarify :)